### DISCUSSION TITLE
Feature request: configurable loggers / json logging

### DISCUSSION THREAD

**Discussion by rburgst** [withastro/roadmap#908] [Proposal]
### Body

# Summary

I would like to be able to get all my logs in json format (node adapter).

# Background & Motivation

When running astro with node in production I want my logs to be json formatted in order to simplify monitoring. Currently 
astro uses a hardcoded `consoleLogDestination` which makes it impossible to get this to work.

# Goals

Being able to provide a custom logger (see pino) would allow me to 

- use any log destination I need
- use any log format I want

**Comment by rburgst** [comment]
the current hardcoded logger is specifically painful for error handling where an error is split up onto multiple lines. 
I have tried packaging my astro server into a fastify server application (as shown in the docs) but I still end up in `App.render` https://github.com/withastro/astro/blob/d7dff648241b23b235f7060249de0fe015411bc3/packages/astro/src/core/app/index.ts#L326
which uses the private hardcoded console logger.
```javascript
    } catch (err) {
      this.#logger.error(null, err.stack || err.message || String(err));
      return this.#renderError(request, { locals, status: 500 });
    }
```
Therefore, even if I created a custom node adapter I wouldnt be able to overwrite the `App.#logger`

**Comment by firxworx** [comment]
+1 x 1000

As Astro adds more back-end features from actions to SSR configurable logging is a must-have feature because it is essential for any real-world business / enterprise applications.

A great example of a business-friendly Node framework that has great support for configurable logging is NestJS.

Check out the docs page for loggers and note how the _expectation_ is that production applications will have specific logging requirements and will replace the built-in logger with their own. This is detailed in the last section "Use external logger": https://docs.nestjs.com/techniques/logger#use-external-logger

_ALL_ logs coming from Astro _must_ be wired through the logger including those from integrations. Just like NestJS, no log entry should be able to "slip by" and be emitted in an invalid format. NO logs by any official code or integration should be emitted using hardcoded `console.log()`.

Consider the common use-case mentioned by the OP for structured logging in JSON format.

Custom loggers need to be supported to meet industry & compliance requirements such as data cleaning / scrubbing of sensitive fields as well as various business requirements.

When this requirement is met, developers can use log data to feed event buses + event streams, it can be dumped into log aggregation services where it can be made searchable / trigger notifications, and it can be dumped into storage buckets where it can be queried with Apache Parquet / AWS Athena. Structured data is easily analyzed by machine learning, etc. Ultimately this will enable Astro-based applications to fully "plug in" to a larger business or enterprise landscape and greater IT architecture.

For the record it is NOT an option to monkey patch `console`!

**Reply by dvelasquez** [reply]
This is exactly what I'm looking for and I need. We are running astro and all the server stdout is sent to logstash and then consumed with kibana.
Having a mix of stdout that is json and other that is not breaks a lot of the functionality there.

For me a good first step compromise would be to be able to setup the log to json and then allow us to provide a logger like winston or pino.

**Comment by ShrJamal** [comment]
I believe it's really important for Astro.js to have this feature, especially for people like me who like to self-host their apps and manage everything themselves. For those using tools like Loki and Grafana to collect and view logs, being able to easily send logs from Astro.js would be super helpful. It would make it much easier for them to keep an eye on things, analyze data, and fine-tune their systems.

**Comment by Eptagone** [comment]
Any updates of this? I also need to customize the default logger.
I want to forward the logs to a custom cloudwatch logger service.

**Comment by JuJup** [comment]
I agree and I found it quite surprising to see that Astro has no default option to add custom loggers or simply to output JSON logs.
It's such a great framework, I'd really appreciate this feature :)

**Comment by ascorbic** [comment]
I've been looking into this. Are people more interested in build logs or SSR/runtime logs? Or both? What about dev?

**Reply by Uninen** [reply]
I'm personally mostly interested in dev logs. Wrote my own Vite plugin that saves both Vite and Browser (console) logs as jsonl but it would be so nice to have this built in to Astro (at least for the non-browser side).

For production this is not that important for me as Sentry covers most needs and the provider logs stuff for me.

**Reply by rburgst** [reply]
I would be interested in runtime logs

**Reply by firxworx** [reply]
runtime logs, observability

While I appreciate the above Sentry comment, not all projects get the benefit of pricey third party solutions. 

I think Astro needs to support structured logging for log aggregation and analysis.

**Comment by ascorbic** [comment]
# Stage 1 RFC

Add support for configurable log handlers in Astro, allowing users to replace the default console output with custom logging implementations (e.g. structured JSON).

## Background & Motivation

Users deploying Astro SSR to production need structured logging for integration with log aggregation services like Kibana, Logstash, CloudWatch, and Grafana/Loki. Agents running builds and dev locally work better with JSON. Today, Astro's logger is hardcoded – `BaseApp` uses `consoleLogDestination` which cannot be replaced, even by adapter authors. Error logs in rendering are split across multiple lines and formatted for human reading, making them unparseable by log aggregation tools.

This is a common requirement for any production SSR deployment and has been widely requested for a long time (this is currently the most upvoted feature request). 

## Goals

- Allow users to configure a custom log handler that receives all Astro log output
- Support structured logging (JSON) for production deployments out of the box
- Provide a `--json` CLI flag for zero-config JSON output (e.g. `astro build --json`, `astro dev --json`)
- Follow the established `{ entrypoint, config }` factory pattern used by session drivers, font providers etc
- Work across all scopes: build, dev, and SSR runtime
- Ensure Vite logs flow through the custom handler (they already route through Astro's logger)
- Ship under `experimental.logger` initially

## Non-Goals

- Request-scoped child loggers with correlation IDs
- Replacing the `debug` package integration – separate concern
- Separate handlers per scope (build vs server) – single handler with platform detection if needed

## Example

Using a built-in handler:

```js
// astro.config.mjs
import { defineConfig, logHandlers } from 'astro/config'

export default defineConfig({
  experimental: {
    logger: logHandlers.json({ pretty: false }),
  },
})
```
A third-party package like `astro-pino-logger` could export a helper function following the same pattern:

```js
import { defineConfig } from 'astro/config'
import { pino } from 'astro-pino-logger'

export default defineConfig({
  experimental: {
    logger: pino({ destination: '/var/log/app.log' }),
    // returns: { entrypoint: 'astro-pino-logger/handler', config: { destination: '/var/log/app.log' } }
  },
})
```

CLI flag (no config needed):

```sh
astro build --json
```

**Reply by matthewp** [reply]
lgtm

**Reply by florian-lefebvre** [reply]
Looks good!

**Reply by ascorbic** [reply]
@ematipico would this work with your refactoring plans?

**Reply by tanepiper** [reply]
👍 This please - I was literally just searching how we can get open telemetry logs from Astro (we have sentry, but would prefer this to go to our otel providers)

**Reply by delucis** [reply]
Hey @tanepiper! We do also have an open proposal for Opent Telemetry as well that might be of interest: https://github.com/withastro/roadmap/issues/1208

**Reply by ematipico** [reply]
Looks good to me too.

**Comment by ematipico** [comment]
This RFC has been accepted and moved to Stage 2 https://github.com/withastro/roadmap/issues/1335
